### PR TITLE
WINDUPRULE-370: resteasy resource method algorithm changes

### PR DESCRIPTION
**Pull request by m-brophy**
https://issues.jboss.org/browse/WINDUPRULE-370

### CODE REVIEW

**Comment by PhilipCattanach** [comment]
I have checked the code and executed the tests. All looks good. I will squash and merge as soon as the automated checks pass. Thanks.

File: rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0"?>
+<ruleset xmlns="http://windup.jboss.org/schema/jboss-ruleset" id="resteasy-resource-method"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://windup.jboss.org/schema/jboss-ruleset http://windup.jboss.org/schema/jboss-ruleset/windup-jboss-ruleset.xsd">
+    <metadata>
+        <dependencies>
+            <addon id="org.jboss.windup.rules,windup-rules-javaee,3.0.0.Final" />
+            <addon id="org.jboss.windup.rules,windup-rules-java,3.0.0.Final" />
+        </dependencies>
+        <targetTechnology id="eap" versionRange="[7,8)" />
```
Comment (mrizzi) [inline rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml]:
```suggestion
        <targetTechnology id="eap" versionRange="[7,8)" />
        <targetTechnology id="resteasy" versionRange="[3,)" />
```

File: rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0"?>
+<ruleset xmlns="http://windup.jboss.org/schema/jboss-ruleset" id="resteasy-resource-method"
```
Comment (mrizzi) [inline rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml]:
```suggestion
<ruleset xmlns="http://windup.jboss.org/schema/jboss-ruleset" id="resteasy30-36"
```

File: rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://windup.jboss.org/schema/jboss-ruleset http://windup.jboss.org/schema/jboss-ruleset/windup-jboss-ruleset.xsd">
+    <metadata>
+        <dependencies>
```
Comment (mrizzi) [inline rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml]:
```suggestion
        <description>
            This ruleset provides analysis of applications that use RESTEasy 3.0 and may require
            individual attention when migrating to RESTEasy 3.6.
        </description>
        <dependencies>
```

File: rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            <addon id="org.jboss.windup.rules,windup-rules-javaee,3.0.0.Final" />
+            <addon id="org.jboss.windup.rules,windup-rules-java,3.0.0.Final" />
+        </dependencies>
+        <targetTechnology id="eap" versionRange="[7,8)" />
+    </metadata>
+    <rules>
+        <!--https://issues.jboss.org/browse/WINDUPRULE-370 -->
+        <rule id="resteasy-resource-method-00001">
```
Comment (mrizzi) [inline rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml]:
```suggestion
        <rule id="resteasy30-36-00001">
```

File: rules-reviewed/eap7/eap71/tests/resteasy-resource-method.windup.test.xml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+<ruletest xmlns="http://windup.jboss.org/schema/jboss-ruleset"
+          id="hibernate51-53-test" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+          xsi:schemaLocation="http://windup.jboss.org/schema/jboss-ruleset http://windup.jboss.org/schema/jboss-ruleset/windup-jboss-ruleset.xsd">
+    <testDataPath>data/data-resteasy-resource-method/</testDataPath>
+    <!--<testDataPath>data/hibernate-core-5.1.6.Final-sources/</testDataPath>-->
+    <rulePath>../resteasy-resource-method.windup.xml</rulePath>
+    <ruleset>
+        <rules>
+            <!--https://issues.jboss.org/browse/WINDUPRULE-370 -->
+            <rule id="resteasy-resource-method-00001-test">
```
Comment (mrizzi) [inline rules-reviewed/eap7/eap71/tests/resteasy-resource-method.windup.test.xml]:
```suggestion
            <rule id="resteasy30-36-00001-test">
```

File: rules-reviewed/eap7/eap71/tests/resteasy-resource-method.windup.test.xml [hunk 6]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0"?>
+<ruletest xmlns="http://windup.jboss.org/schema/jboss-ruleset"
+          id="hibernate51-53-test" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (mrizzi) [inline rules-reviewed/eap7/eap71/tests/resteasy-resource-method.windup.test.xml]:
```suggestion
          id="resteasy30-36-test" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```

File: rules-reviewed/eap7/eap71/tests/resteasy-resource-method.windup.test.xml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+          xsi:schemaLocation="http://windup.jboss.org/schema/jboss-ruleset http://windup.jboss.org/schema/jboss-ruleset/windup-jboss-ruleset.xsd">
+    <testDataPath>data/data-resteasy-resource-method/</testDataPath>
+    <!--<testDataPath>data/hibernate-core-5.1.6.Final-sources/</testDataPath>-->
```
Comment (mrizzi) [inline rules-reviewed/eap7/eap71/tests/resteasy-resource-method.windup.test.xml]:
```suggestion
```

File: rules-reviewed/eap7/eap71/tests/resteasy-resource-method.windup.test.xml [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    <rulePath>../resteasy-resource-method.windup.xml</rulePath>
```
Comment (mrizzi) [inline rules-reviewed/eap7/eap71/tests/resteasy-resource-method.windup.test.xml]:
```suggestion
    <rulePath>../resteasy30-36.windup.xml</rulePath>
```

File: rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+            <when>
+                <project>
+                    <artifact groupId="org.jboss.spec.javax.ws.rs" artifactId="jboss-jaxrs-api_2.0_spec" />
+                </project>
+            </when>
+            <perform>
+                <hint title="RESTEasy Resource Method Algorithm Switch changes" effort="1" category-id="potential">
```
Comment (mrizzi) [inline rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml]:
```suggestion
                <hint title="RESTEasy 3.6 - Resource Method Algorithm Switch changes" effort="1" category-id="potential">
```

File: rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+                    <message>If you migrate your application from JBoss EAP from 7.1.1 to 7.2.x and want to retain the stricter behavior as defined in the JAX-RS 2.0 specification, set the `jaxrs.2.0.request.matching` option to true.</message>
```
Comment (mrizzi) [inline rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml]:
```suggestion
                    <message>If you migrate your application from JBoss EAP from 7.1.1 to 7.2.x and want to retain the stricter behavior as defined in the JAX-RS 2.0 specification, set the `jaxrs.2.0.request.matching` option to `true`.</message>
```

Comment (mrizzi) [reply rules-reviewed/eap7/eap71/resteasy-resource-method.windup.xml]: this has to be applied with the ` char and not the ' one :wink:

File: rules-reviewed/eap7/eap71/resteasy30-36.windup.xml [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        <!--https://issues.jboss.org/browse/WINDUPRULE-370 -->
+        <rule id="resteasy30-36-00001">
+            <when>
+                <project>
+                    <artifact groupId="org.jboss.spec.javax.ws.rs" artifactId="jboss-jaxrs-api_2.0_spec" />
+                </project>
+            </when>
+            <perform>
+                <hint title="RESTEasy 3.6 - Resource Method Algorithm Switch changes" effort="1" category-id="potential">
+                    <message>If you migrate your application from JBoss EAP from 7.1.1 to 7.2.x and want to retain the stricter behavior as defined in the JAX-RS 2.0 specification, set the `jaxrs.2.0.request.matching` option to 'true'.</message>
```
Comment (mrizzi) [inline rules-reviewed/eap7/eap71/resteasy30-36.windup.xml]: Sorry to still bother you with this but please apply this ` char to have the right formatting.
```suggestion
                    <message>If you migrate your application from JBoss EAP from 7.1.1 to 7.2.x and want to retain the stricter behavior as defined in the JAX-RS 2.0 specification, set the `jaxrs.2.0.request.matching` option to `true`.</message>
```

**Comment by PhilipCattanach** [comment]
I can see the resteasy targetTechnology tag in the rule and the appropriate characters to highlight the property and value in the message. The Test runs successfully, the appropriate words highlighted, and the link takes you to the relevant part of the docs. All good.

**Pull request merged**